src/upload-helper.ts

```diff
@@ -56,58 +61,64 @@ export class UploadHelper {
       predefinedAcl?: PredefinedAcl;
     }
     const options: UploadOptions = { gzip, predefinedAcl };
+    const normalizedFilePath = path.posix.normalize(filename);
+    // set destination if defined
     if (destination) {
-      // If obj prefix is set, then extract filename and append to prefix.
-      options.destination = `${destination}/${path.posix.basename(filename)}`;
+      options.destination = destination;
+    }
+    if (!process.env.UPLOAD_ACTION_NO_LOG) {
+      core.info(
+        `Uploading file: ${normalizedFilePath} to gs://${bucketName}/${
+          destination ? destination : normalizedFilePath
+        }`,
+      );
     }
     const uploadedFile = await this.storage
       .bucket(bucketName)
-      .upload(filename, options);
+      .upload(normalizedFilePath, options);
     return uploadedFile;
   }
 
   /**
    * Uploads a specified directory to a GCS bucket. Based on
-   * https://github.com/googleapis/nodejs-storage/blob/master/samples/uploadDirectory.js
    *
    * @param bucketName The name of the bucket.
    * @param directoryPath The path of the directory to upload.
-   * @param objectKeyPrefix Optional Prefix for in the GCS bucket.
-   * @param clearExistingFilesFirst Clean files in the prefix before uploading.
+   * @param glob Glob pattern if any.
+   * @param gzip Flag to enable gzip.
+   * @param parent Flag to enable parent dir in destination path.
+   * @param predefinedAcl Predefined ACL config.
+   * @param concurrency Number of files simultaneously uploaded.
    * @returns The list of UploadResponses which contains the file and metadata.
    */
   async uploadDirectory(
     bucketName: string,
     directoryPath: string,
+    glob = '',
     gzip: boolean,
     prefix = '',
+    parent = true,
     predefinedAcl?: PredefinedAcl,
+    concurrency = 100,
   ): Promise<UploadResponse[]> {
-    const pathDirName = path.posix.dirname(directoryPath);
-    // Get list of files in the directory.
-    const filesList = await getFiles(directoryPath);
-
-    const resp = await Promise.all(
-      filesList.map(async (filePath) => {
-        // Get relative path from directoryPath.
-        let destination = `${path.posix.dirname(
-          path.posix.relative(pathDirName, filePath),
-        )}`;
-        // If prefix is set, prepend.
-        if (prefix) {
-          destination = `${prefix}/${destination}`;
-        }
-
-        const uploadResp = await this.uploadFile(
-          bucketName,
-          filePath,
-          gzip,
-          destination,
-          predefinedAcl,
-        );
-        return uploadResp;
-      }),
-    );
-    return resp;
+    // const filesList = await getFiles(directoryPath,parent);
+    const filesList = await globby([path.posix.join(directoryPath, glob)]);
```

Comment:
nit: can you add a comment here. I assume if glob = '' it returns all files?

Reply:
added a comment, `glob = ''` evaluates `path.posix.join(directoryPath, glob)` to just `directoryPath/` which then globby uses as `directoryPath/**/*`. This is default behavior where previously we were recursively walking the dir.